File: interview_questions/placement_practise/src/linkedlist/linkedlist.java
package linkedlist;

import static linkedlist.linkedlist.node.del;
import static linkedlist.linkedlist.node.insert;

public class linkedlist
{
    node head;
    static  class node
    {

        int data;
        node next;

        node(int k){
            data=k;
            next=null;

        }


        public static linkedlist insert(linkedlist n, int data){
            node new_node=new node(data);

            if(n.head==null){
                n.head=new_node;
            }

            else{
                node temp=n.head;
                while(temp.next!=null){
                    temp=temp.next;
                }

                temp.next=new_node;
            }
            return n;
        }

        public static void print(linkedlist n){
            node temp=n.head;
            while(temp.next!=null){
                System.out.print(temp.data+"-->");
                temp=temp.next;
            }
            System.out.print("null");


        }

        public static void del(linkedlist n, int d){
            node temp=n.head;

            if(temp.data==d){
                n.head=temp.next;
                return;
            }

            while (temp.next.data!=d){
                temp=temp.next;
            }


            node todel=temp.next;
            temp.next=temp.next.next;


        }

        public static void mid(linkedlist n)
        {
            node fast=n.head;
            node slow=n.head;

            while(fast!=null && fast.next!=null)
            {
                fast=fast.next.next;
                slow=slow.next;
            }

            System.out.println(slow.data);
        }
    }



    public static void main(String[] args) {
        linkedlist list=new linkedlist();

        list=insert(list,3);
        list=insert(list,4);
        list=insert(list,2);
        list=insert(list,9);
        node.print(list);
        node.mid(list);
        node.print(list);
    }

}

File: interview_questions/linked list/linkedlist.cpp
#include <iostream>
using namespace std;

class node
{
public:
    int data;
    node *next;

    node(int val)
    {
        data = val;
        next = NULL;
    }
};

void insert(node *&head, int m)
{

    node *n = new node(m);
    if (head == NULL)
    {
        head = n;
        return;
    }

    node *temp = head;

    while (temp->next != NULL)
    {
        temp = temp->next;
    }

    temp->next = n;
}

void display(node *&head)
{
    node *temp = head;
    while (temp != NULL)
    {
        cout << temp->data << "-->";
        temp = temp->next;
    }
    cout << "NULL" << endl;
}

void del(node *&head, int del)
{
    node *temp = head;
    while (temp->next->data != del)
    {
        temp = temp->next;
    }

    node *todel = temp->next;

    temp->next = temp->next->next;
    delete todel;
}

void middle(node *&head)
{
    node *fst = head;
    node *slw = head;

    while (fst != NULL && fst->next != NULL)
    {
        fst = fst->next->next;
        slw = slw->next;
    }

    cout << slw->data << endl;
}

void reverse(node *&head)
{
    node *curr = head;
    node *nxt = NULL;
    node *prev = NULL;

    while (curr != NULL)
    {
        nxt = curr->next;
        curr->next = prev;
        prev = curr;
        curr = nxt;
    }

    head = prev;
}

bool detectloop(node *&head)
{
    node *fst = head;
    node *slw = head;

    while (fst && slw && fst->next)
    {
        if (slw == fst)
        {
            return true;
        }
    }

    return false;
}
int main()

{

    node *head = NULL;

    insert(head, 8);
    insert(head, 7);
    insert(head, 10);
    insert(head, 3);
    del(head, 7);
    display(head);
    middle(head);

    reverse(head);
    display(head);

    head->next->next = head;
    cout << detectloop(head);
}